### ISSUE TITLE
subprocess causes errors in asyncio.subprocess 

### ISSUE THREAD

**Issue by wouterdb** [pytest-dev/pytest-asyncio#114]
# TLDR;

when combining async and sync testcases that use subprocesses, the IOloop childwatcher is not correctly closed. 

workaround included

# How to reproduce

```python
import pytest
import asyncio
import subprocess


@pytest.mark.asyncio
async def test_setup(event_loop):
    # run a process using asyncio.subprocess
    process = await asyncio.subprocess.create_subprocess_exec("/usr/bin/echo", "test", stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    out, err = await asyncio.wait_for(process.communicate(), timeout=30)


def test_sync(capsys):
   # generate some signals by running LOTS of processes 
    for i in range(1000):
        process = subprocess.check_output(["/usr/bin/echo", "test"], stderr=subprocess.STDOUT)

        out, err = capsys.readouterr()
        assert len(err) == 0, err
        print(out, process)
```

the second testcase fails if it is run after the first one

```
―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_sync ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

capsys = <_pytest.capture.CaptureFixture object at 0x7fd991ce7b38>

    def test_sync(capsys):
        for i in range(1000):
            process = subprocess.check_output(["/usr/bin/echo", "test"], stderr=subprocess.STDOUT)
    
            out, err = capsys.readouterr()
>           assert len(err) == 0, "got unexpected errors on iterations %d\n%s"%(i, err)
E           AssertionError: got unexpected errors on iterations 278
E             Exception ignored when trying to write to the signal wakeup fd:
E             BlockingIOError: [Errno 11] Resource temporarily unavailable
E             
E           assert 125 == 0
E             -125
E             +0

test_test.py:17: AssertionError
```

# Reason

When using asyncio.subprocess, a [self-pipe](https://github.com/python/cpython/blob/fb2c3465f09e1f720cdae7eca87d62125a327fd9/Lib/asyncio/selector_events.py#L55) is created. The python interpreter [writes to this pipe whenever a signal arrives](https://github.com/python/cpython/blob/fdf282d609fd172d52b59a6f1f062eb701494528/Modules/signalmodule.c#L237). 

When the pipe is not read (e.g. when the ioloop is not running), it fills up, when it is full and new signals arrive, the interpreter panics.

If the IOloop is stopped, the pipe is detached. [However, when the IOloop is replaced, the childwatcher is moved to the new ioloop](https://github.com/python/cpython/blob/fb2c3465f09e1f720cdae7eca87d62125a327fd9/Lib/asyncio/unix_events.py#L1116). The new IOloop will then attach a new pipe.

Before pytest_asyncio stops the ioloop, [it first reinstalls the old ioloop](https://github.com/pytest-dev/pytest-asyncio/blob/master/pytest_asyncio/plugin.py#L138). 
This causes the self-pipe to remain attached and fill up with junk, causing the interperter to panic.

# Workaround

```python
@pytest.fixture(scope="function", autouse=True)
def reset_all():
   asyncio.set_child_watcher(None)
```
# Solution
I think the child_watcher reset could also be included in the pytest_asyncio fixtures as well

**Comment by asvetlov** [comment]
Python 3.8+ doesn't have this problem with the default configuration.
For 3.7 we can consider backporting of `asyncio.ThreadedChildWatcher`, a champion is welcome!

**Comment by seifertm** [comment]
I can confirm that the provided example fails on CPython 3.7, but succeeds with CPython>=3.8 (tested on Linux x86_64).

CPython 3.7 will be end of life roughly 9 months from now (see [PEP 537](https://peps.python.org/pep-0537/)). Child watchers are currently being deprecated (see https://github.com/python/cpython/issues/94597) and are scheduled for removal in Python 3.14.

If you are affected by this issue, your best course of action will likely be upgrading to a new Python version.

I'll close this issue as "won't fix" since the pytest-asyncio maintainers are unlikely to implement the backport of said child watcher. If anyone has a different opinion on the matter, please reopen the issue. We'll consider pull requests.
